File: app/src/main/java/com/iita/akilimo/views/fragments/dialog/FertilizerPriceDialogFragment.java
package com.iita.akilimo.views.fragments.dialog;


import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;

import com.crashlytics.android.Crashlytics;
import com.google.android.gms.common.util.Strings;
import com.google.android.material.textfield.TextInputLayout;
import com.iita.akilimo.R;
import com.iita.akilimo.interfaces.IDismissListener;
import com.iita.akilimo.models.Fertilizer;
import com.iita.akilimo.models.FertilizerPrices;
import com.iita.akilimo.utils.MathHelper;
import com.iita.akilimo.utils.objectbox.ObjectBoxEntityProcessor;

import java.util.List;

import butterknife.ButterKnife;

/**
 * A simple {@link androidx.fragment.app.Fragment} subclass.
 */
public class FertilizerPriceDialogFragment extends DialogFragment {

    private static final String LOG_TAG = FertilizerPriceDialogFragment.class.getSimpleName();

    public static final String FERTILIZER_TYPE = "selected_type";
    public static final String ARG_ITEM_ID = "fertilizer_dialog_fragment";

    private boolean isExactPriceRequired = false;
    private boolean isPriceValid = false;
    private boolean priceSpecified = false;
    private boolean removeSelected = false;

    private Dialog dialog;
    private RadioGroup radioGroup;
    private TextInputLayout exactPriceWrapper;
    private EditText editExactFertilizerPrice;

    private Button btnClose;
    private Button btnUpdate;
    private Button btnRemove;


    private MathHelper mathHelper;
    private Context context;
    private ObjectBoxEntityProcessor objectBox;
    private Fertilizer fertilizer;
    private List<FertilizerPrices> fertilizerPricesList;

    private double savedPricePerBag = 0.0;
    private String countryCode;
    private String currencyCode;
    private String bagPrice;
    private String bagPriceRange = "NA";
    private String exactPrice = "0";

    private IDismissListener onDismissListener;

    public FertilizerPriceDialogFragment() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        this.context = context;
        objectBox = ObjectBoxEntityProcessor.getInstance(context);
        mathHelper = new MathHelper();
    }


    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {

        Bundle bundle = this.getArguments();
        if (bundle != null) {
            fertilizer = bundle.getParcelable(FERTILIZER_TYPE);
        }
        dialog = new Dialog(context);

        dialog.getWindow().requestFeature(Window.FEATURE_NO_TITLE);
        dialog.getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN);
        dialog.setContentView(R.layout.fragment_fertilizer_price_dialog);
        ButterKnife.bind(dialog);

        dialog.setCancelable(true);
        dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        dialog.getWindow().getAttributes().windowAnimations = R.style.DialogAnimation;


        btnClose = dialog.findViewById(R.id.close_button);
        btnUpdate = dialog.findViewById(R.id.update_button);
        btnRemove = dialog.findViewById(R.id.remove_button);

        radioGroup = dialog.findViewById(R.id.radioGroup);
        exactPriceWrapper = dialog.findViewById(R.id.exactPriceWrapper);
        editExactFertilizerPrice = dialog.findViewById(R.id.editExactFertilizerPrice);
        TextView lblPricePerBag = dialog.findViewById(R.id.lblFragmentTitle);

        if (fertilizer != null) {
            countryCode = fertilizer.getCountryCode();
            currencyCode = fertilizer.getCurrency();
            String titleText = context.getString(R.string.price_per_bag, currencyCode, fertilizer.getName());
            lblPricePerBag.setText(titleText);
        }


        btnClose.setOnClickListener(view -> {
            priceSpecified = false;
            removeSelected = false;
            dismiss();
        });

        btnRemove.setOnClickListener(view -> {
            priceSpecified = false;
            removeSelected = true;
            fertilizer.setPrice("NA");
            fertilizer.setPricePerBag(0.0);
            fertilizer.setPriceRange(null);
            fertilizer.setSelected(false);
            fertilizer.setExactPrice(false);
            dismiss();
        });
        //save the data
        btnUpdate.setOnClickListener(v -> {
            if (isExactPriceRequired) {
                bagPrice = editExactFertilizerPrice.getText().toString();
                if (Strings.isEmptyOrWhitespace(bagPrice)) {
                    editExactFertilizerPrice.setError("Please provide a valid bag price");
                    isPriceValid = false;
                    return;
                }
                savedPricePerBag = Double.parseDouble(bagPrice);
                bagPriceRange = mathHelper.formatNumber(savedPricePerBag, currencyCode);
                isPriceValid = true;
                editExactFertilizerPrice.setError(null);
            }
            fertilizer.setPrice(bagPrice);
            fertilizer.setPricePerBag(savedPricePerBag);
            fertilizer.setPriceRange(bagPriceRange);
            fertilizer.setSelected(true);
            fertilizer.setExactPrice(isExactPriceRequired);

            if (isPriceValid) {
                priceSpecified = true;
                removeSelected = false;
                dismiss();
            }

        });

        radioGroup.setOnCheckedChangeListener((radioGroup, i) -> radioSelected(radioGroup));
        if (objectBox != null) {
            fertilizerPricesList = objectBox.getFertilizerPrices(countryCode);
            addPriceRadioButtons(fertilizerPricesList, fertilizer);
        }
        return dialog;
    }

    private void radioSelected(RadioGroup radioGroup) {
        int radioButtonId = radioGroup.getCheckedRadioButtonId();
        RadioButton radioButton = dialog.findViewById(radioButtonId);
        long itemTagIndex = (long) radioButton.getTag();

        try {
            FertilizerPrices pricesResp = fertilizerPricesList.get((int) itemTagIndex);
            isExactPriceRequired = false;
            isPriceValid = true;
            savedPricePerBag = pricesResp.getPricePerBag();
            bagPrice = String.valueOf(savedPricePerBag);
            bagPriceRange = pricesResp.getPriceRange();


            exactPriceWrapper.setVisibility(View.GONE);
            exactPriceWrapper.getEditText().setText(null);
            if (savedPricePerBag >= 0 && savedPricePerBag <= 0) {
                bagPrice = "NA";
                bagPriceRange = "NA";
            } else if (savedPricePerBag < 0) {
                isExactPriceRequired = true;
                isPriceValid = false;
                exactPriceWrapper.setVisibility(View.VISIBLE);
            }
        } catch (Exception ex) {
            Crashlytics.log(Log.ERROR, LOG_TAG, "Radio selection issues");
            Crashlytics.logException(ex);
        }
    }

    private void addPriceRadioButtons(List<FertilizerPrices> fertilizerPricesList, Fertilizer fertilizer) {
        radioGroup.removeAllViews();
        double selectedPrice = 0.0;
        if (fertilizer != null) {
            selectedPrice = fertilizer.getPricePerBag();
            isExactPriceRequired = fertilizer.isExactPrice();
            if (fertilizer.isSelected()) {
                btnUpdate.setText(R.string.lbl_update);
                btnRemove.setVisibility(View.VISIBLE);
            }
            if (isExactPriceRequired) {
                exactPrice = fertilizer.getPrice();
            }
        }
        for (FertilizerPrices pricesResp : fertilizerPricesList) {

            long listIndex = pricesResp.getPriceId() - 1;//reduce by one so as to match the index in the list

            RadioButton radioButton = new RadioButton(getActivity());
            radioButton.setId(View.generateViewId());
            radioButton.setTag(listIndex);
//            radioButton.setTextSize(TypedValue.COMPLEX_UNIT_PX, getResources().getDimension(R.dimen.spacing_large));

            double price = pricesResp.getPricePerBag();
            String radioLabel = pricesResp.getPriceRange();
            if (price >= 0 && price <= 0) {
                radioLabel = context.getString(R.string.lbl_do_not_know);
            } else if (price < 0) {
                radioLabel = context.getString(R.string.exact_fertilizer_price);
            }
            radioButton.setText(radioLabel);

            radioGroup.addView(radioButton);
            //set relevant radio button as selected based on the price range
            if (isExactPriceRequired) {
                radioButton.setChecked(true);
                isPriceValid = true;
                exactPriceWrapper.setVisibility(View.VISIBLE);
                editExactFertilizerPrice.setText(exactPrice);
            } else if (pricesResp.getPricePerBag() == selectedPrice) {
                radioButton.setChecked(true);
            }
        }
    }

    @Override
    public void onDismiss(@NonNull DialogInterface dialog) {
        super.onDismiss(dialog);
        if (onDismissListener != null) {
            onDismissListener.onDismiss(this.priceSpecified, this.fertilizer, this.removeSelected);
        }
    }

    public void setOnDismissListener(IDismissListener dismissListener) {
        this.onDismissListener = dismissListener;
    }
}

File: app/src/main/java/com/akilimo/mobile/views/fragments/dialog/FertilizerPriceDialogFragment.kt
package com.akilimo.mobile.views.fragments.dialog

import android.app.Dialog
import android.content.DialogInterface
import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.view.Window
import android.view.WindowManager
import android.widget.RadioButton
import android.widget.RadioGroup
import androidx.core.graphics.drawable.toDrawable
import com.akilimo.mobile.R
import com.akilimo.mobile.databinding.FragmentFertilizerPriceDialogBinding
import com.akilimo.mobile.entities.Fertilizer
import com.akilimo.mobile.entities.FertilizerPrice
import com.akilimo.mobile.inherit.BaseDialogFragment
import com.akilimo.mobile.interfaces.IFertilizerDismissListener
import com.akilimo.mobile.utils.CurrencyCode
import com.akilimo.mobile.utils.enums.EnumCountry
import io.sentry.Sentry

class FertilizerPriceDialogFragment : BaseDialogFragment() {
    private var isExactPriceRequired = false
    private var isPriceValid = false
    private var priceSpecified = false
    private var removeSelected = false

    private var dialog: Dialog? = null
    private var fertilizer: Fertilizer? = null
    private var fertilizerPricesList: List<FertilizerPrice>? = null

    private var savedPricePerBag = 0.0
    private var maxPrice = 0.0
    private var minPrice = 0.0
    private var countryCode: String? = null
    private var currencyCode: String? = null
    private var fertilizerKey: String? = null
    private var currencyName: String? = null
    private var bagPrice: Double? = null
    private var bagPriceRange: String? = "NA"
    private var exactPrice = 0.0

    private var onDismissListener: IFertilizerDismissListener? = null
    private var _binding: FragmentFertilizerPriceDialogBinding? = null
    private val binding get() = _binding!!

    companion object {
        private val LOG_TAG: String = FertilizerPriceDialogFragment::class.java.simpleName
        const val FERTILIZER_TYPE: String = "selected_type"
        const val ARG_ITEM_ID: String = "fertilizer_dialog_fragment"
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val bundle = arguments
        val context = requireContext()
        if (bundle != null) {
            fertilizer = bundle.getParcelable(FERTILIZER_TYPE)
        }
        dialog = Dialog(context)
        _binding = FragmentFertilizerPriceDialogBinding.inflate(layoutInflater)
        dialog!!.setContentView(binding.root)

        dialog!!.window!!.apply {
            requestFeature(Window.FEATURE_NO_TITLE)
            setFlags(
                WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN
            )
            setBackgroundDrawable(Color.TRANSPARENT.toDrawable())
            attributes.windowAnimations = R.style.DialogSlideAnimation
        }

        if (fertilizer != null) {
            countryCode = fertilizer!!.countryCode
            currencyCode = fertilizer!!.currencyCode
            fertilizerKey = fertilizer!!.fertilizerKey
            currencySymbol = currencyCode
            val extendedCurrency = CurrencyCode.getCurrencySymbol(currencyCode)
            if (extendedCurrency != null) {
                currencySymbol = extendedCurrency.symbol
                currencyName = extendedCurrency.name
            }
            currencyName = when (countryCode) {
                "TZ" -> EnumCountry.Tanzania.currencyName(context)
                "NG" -> EnumCountry.Nigeria.currencyName(context)
                "GH" -> EnumCountry.Ghana.currencyName(context)
                "RW" -> EnumCountry.Rwanda.currencyName(context)
                "BI" -> EnumCountry.Burundi.currencyName(context)
                else -> currencyName
            }
            val titleText =
                context.getString(R.string.price_per_bag, fertilizer!!.name, currencyName)
            binding.lblFragmentTitle.text = titleText
        }

        binding.closeButton.setOnClickListener {
            priceSpecified = false
            removeSelected = false
            dismiss()
        }

        binding.removeButton.setOnClickListener {
            priceSpecified = false
            removeSelected = true
            fertilizer?.apply {
                price = 0.0
                pricePerBag = 0.0
                priceRange = null
                selected = false
                exactPrice = false
            }
            dismiss()
        }

        binding.updateButton.setOnClickListener {
            if (isExactPriceRequired) {
                try {
                    bagPrice = binding.editExactFertilizerPrice.text.toString().toDouble()
                } catch (ex: Exception) {
                    Sentry.captureException(ex)
                    bagPrice = 0.0
                }
                savedPricePerBag = bagPrice!!
                bagPriceRange = mathHelper!!.formatNumber(savedPricePerBag, currencyCode)
                isPriceValid = true
                binding.editExactFertilizerPrice.error = null
            }
            fertilizer?.apply {
                price = bagPrice!!
                pricePerBag = savedPricePerBag
                priceRange = bagPriceRange
                selected = true
                exactPrice = isExactPriceRequired
            }
            if (isPriceValid) {
                priceSpecified = true
                removeSelected = false
                dismiss()
            }
        }

        binding.radioGroup.setOnCheckedChangeListener { group, _ ->
            radioSelected(group)
        }

        if (database != null) {
            fertilizerPricesList =
                database!!.fertilizerPriceDao().findAllByFertilizerKey(fertilizerKey!!)
            addPriceRadioButtons(fertilizerPricesList!!, fertilizer)
        }

        return dialog!!
    }

    private fun radioSelected(radioGroup: RadioGroup) {
        val radioButtonId = radioGroup.checkedRadioButtonId
        val radioButton = dialog!!.findViewById<RadioButton>(radioButtonId)
        val itemTagIndex = radioButton.tag as Long

        currencySymbol = currencyCode
        CurrencyCode.getCurrencySymbol(currencyCode)?.let {
            currencySymbol = it.symbol
        }

        try {
            val pricesResp = database!!.fertilizerPriceDao().findOneByPriceId(itemTagIndex)
            isExactPriceRequired = false
            isPriceValid = true
            savedPricePerBag = pricesResp.pricePerBag
            bagPriceRange = pricesResp.priceRange
            bagPrice = savedPricePerBag
            binding.exactPriceWrapper.visibility = View.GONE
            binding.exactPriceWrapper.editText?.text = null

            if (savedPricePerBag == 0.0) {
                bagPrice = 0.0
                bagPriceRange = getString(R.string.lbl_unknown_price)
            } else if (savedPricePerBag < 0) {
                isExactPriceRequired = true
                isPriceValid = false
                binding.exactPriceWrapper.visibility = View.VISIBLE
            }
        } catch (ex: Exception) {
            Sentry.captureException(ex)
        }
    }

    private fun addPriceRadioButtons(
        fertilizerPricesList: List<FertilizerPrice>,
        fertilizer: Fertilizer?
    ) {
        binding.radioGroup.removeAllViews()
        val context = requireContext()
        var selectedPrice = 0.0
        currencySymbol = currencyCode
        CurrencyCode.getCurrencySymbol(currencyCode)?.let {
            currencySymbol = it.symbol
        }

        fertilizer?.let {
            selectedPrice = it.pricePerBag
            isExactPriceRequired = it.exactPrice
            if (it.selected) {
                binding.updateButton.setText(R.string.lbl_update)
                binding.removeButton.visibility = View.VISIBLE
            }
            if (isExactPriceRequired) {
                isPriceValid = true
                exactPrice = it.price
                if (exactPrice < 0) exactPrice = 0.0
                binding.editExactFertilizerPrice.setText(exactPrice.toString())
            }
        }

        for ((_, _, priceId, _, _, _, minAllowedPrice, maxAllowedPrice, price, _, priceRange) in fertilizerPricesList) {
            minPrice = minAllowedPrice
            maxPrice = maxAllowedPrice
            val listIndex = priceId.toLong()
            val radioButton = RadioButton(activity).apply {
                id = View.generateViewId()
                tag = listIndex
                text = when {
                    price == 0.0 -> context.getString(R.string.lbl_do_not_know)
                    price < 0 -> {
                        binding.exactPriceWrapper.hint =
                            getString(R.string.exact_fertilizer_price_currency, currencyName)
                        context.getString(R.string.exact_fertilizer_price)
                    }

                    else -> context.getString(R.string.lbl_about, priceRange)
                }
            }
            binding.radioGroup.addView(radioButton)

            if (isExactPriceRequired) {
                radioButton.isChecked = true
                isPriceValid = true
                binding.exactPriceWrapper.visibility = View.VISIBLE
                binding.editExactFertilizerPrice.setText(exactPrice.toString())
            } else if (price == selectedPrice) {
                radioButton.isChecked = true
            }
            isExactPriceRequired = false
        }
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        _binding = null
        onDismissListener?.onDismiss(priceSpecified, fertilizer!!, removeSelected)
    }

    fun setOnDismissListener(dismissListener: IFertilizerDismissListener?) {
        this.onDismissListener = dismissListener
    }
}
